pcsx2/gui/Dialogs/GSDumpDialog.cpp

```diff
@@ -0,0 +1,768 @@
+				GSData data = {id, (char*)vsync, 1, Dummy};
+				m_root_window->m_dump_packets.push_back(data);
+				break;
+			}
+			case ReadFIFO2:
+			{
+				u32 fifo = 0;
+				dump_file.Read(&fifo, 4);
+				GSData data = {id, (char*)fifo, 4, Dummy};
+				m_root_window->m_dump_packets.push_back(data);
+				break;
+			}
+			case Registers:
+			{
+				char regs_tmp[8192];
+				dump_file.Read(&regs, 8192);
+				GSData data = {id, regs_tmp, 8192, Dummy};
+				m_root_window->m_dump_packets.push_back(data);
+				break;
+			}
+		}
+	}
+
+	if (m_root_window->m_debug_mode->GetValue())
+		m_root_window->GenPacketList(m_root_window->m_dump_packets);
+
+	//return here to debug pre gs
+	//return;
+
+	GetCorePlugins().Init();
+	GSsetBaseMem((void*)regs);
+	if (GSopen2((void*)pDsp, renderer_override) != 0)
+	{
+		OnStop();
+		return;
+	}
+
+	GSsetGameCRC((int)crc, 0);
+
+	if (!GetCorePlugins().DoFreeze(PluginId_GS, 0, &fd, true))
+		m_running = false;
+	GSvsync(1);
+	GSreset();
+	GSsetBaseMem((void*)regs);
+	GetCorePlugins().DoFreeze(PluginId_GS, 0, &fd, true);
+
+	size_t i = 0;
+	size_t RunTo = 0;
+	size_t debug_idx = 0;
+
+	while (m_running)
+	{
+		if (m_root_window->m_debug_mode->GetValue())
+		{
+			if (m_root_window->m_button_events.size() > 0)
+			{
+				switch (m_root_window->m_button_events[0].index)
+				{
+					case Step:
+						if (debug_idx >= m_root_window->m_dump_packets.size())
+							debug_idx = 0;
+						RunTo = debug_idx;
+						break;
+					case RunCursor:
+						RunTo = m_root_window->m_button_events[0].index;
+						if (debug_idx > RunTo)
+							debug_idx = 0;
+						break;
+					case RunVSync:
+						if (debug_idx >= m_root_window->m_dump_packets.size())
+							debug_idx = 1;
+						auto it = std::find_if(m_root_window->m_dump_packets.begin() + debug_idx, m_root_window->m_dump_packets.end(), [](const GSData& gs) { return gs.id == Registers; });
+						if (it != std::end(m_root_window->m_dump_packets))
+							RunTo = std::distance(m_root_window->m_dump_packets.begin(), it);
+						break;
+				}
+				m_root_window->m_button_events.erase(m_root_window->m_button_events.begin());
+
+				if (debug_idx <= RunTo)
+				{
+					while (debug_idx <= RunTo)
+					{
+						m_root_window->ProcessDumpEvent(m_root_window->m_dump_packets[debug_idx++], regs);
+					}
+					auto it = std::find_if(m_root_window->m_dump_packets.begin() + debug_idx, m_root_window->m_dump_packets.end(), [](const GSData& gs) { return gs.id == Registers; });
+					if (it != std::end(m_root_window->m_dump_packets))
+						m_root_window->ProcessDumpEvent(*it, regs);
+
+					debug_idx--;
+				}
+
+				// do vsync
+				m_root_window->ProcessDumpEvent(GSData{VSync, 0, 0, Dummy}, regs);
+			}
+		}
+		else
+		{
+			while (i < (m_root_window->m_dump_packets.size()-1))
+			{
+				m_root_window->ProcessDumpEvent(m_root_window->m_dump_packets[i++], regs);
+
+				if (m_root_window->m_dump_packets[i].id == VSync)
+					break;
+			}
+			if (i >= m_root_window->m_dump_packets.size())
+				i = 0;
```

Comment:
think about this code with me:

while loop: you compare `i < size -1`
if statement: you compare `i >= size`

so what if `i ==  size - 1`?

Am I misunderstanding something or did you just write an infinite loop?

Reply:
The original code did that too, am I misunderstanding something wrt the inner workings of this loop?
```csharp
                                while (gs_idx < dump.Data.Count)
                                {
                                    GSData itm = dump.Data[gs_idx++];
                                    CurrentGIFPacket = itm;
                                    Step(itm, pointer);

                                    if (gs_idx < dump.Data.Count && dump.Data[gs_idx].id == GSType.VSync)
                                        break;
                                }

                                if (gs_idx >= dump.Data.Count) gs_idx = 0;
```

Reply:
that code doesn't subtract by 1 in one comparison and not in the other.